src/bijectors/coupling.jl

```diff
@@ -0,0 +1,273 @@
+@inline partition(m::PartitionMask, x) = (transpose(m.A_1) * x, transpose(m.A_2) * x, transpose(m.A_3) * x)
+
+
+# Coupling
+
+"""
+    Coupling{F, M}(θ::F, mask::M)
+
+Implements a coupling-layer as defined in [1].
+
+# Examples
+```julia-repl
+julia> m = PartitionMask(3, [1], [2]) # <= going to use x[2] to parameterize transform of x[1]
+PartitionMask{SparseArrays.SparseMatrixCSC{Float64,Int64}}(
+  [1, 1]  =  1.0, 
+  [2, 1]  =  1.0, 
+  [3, 1]  =  1.0)
+
+julia> cl = Coupling(θ -> Shift(θ[1]), m) # <= will do `y[1:1] = x[1:1] + x[2:2]`;
+
+julia> x = [1., 2., 3.];
+
+julia> cl(x)
+3-element Array{Float64,1}:
+ 3.0
+ 2.0
+ 3.0
+
+julia> inv(cl)(cl(x))
+3-element Array{Float64,1}:
+ 1.0
+ 2.0
+ 3.0
+
+julia> coupling(cl) # get the `Bijector` map `θ -> b(⋅, θ)`
+Shift
+
+julia> couple(cl, x) # get the `Bijector` resulting from `x`
+Shift{Array{Float64,1},1}([2.0])
+```
+
+# References
+[1] Kobyzev, I., Prince, S., & Brubaker, M. A., Normalizing flows: introduction and ideas, CoRR, (),  (2019). 
+"""
+struct Coupling{F, M} <: Bijector{1} where {F, M <: PartitionMask}
+    θ::F
+    mask::M
+end
+
+function Coupling(θ, n::Int)
+    idx = Int(floor(n / 2))
+    return Coupling(θ, PartitionMask(n, 1:idx))
+end
+
+function Coupling(cl::Coupling{B}, mask::PartitionMask) where {B}
+    return Coupling(cl.θ, mask)
+end
+
+"Returns the constructor of the coupling law."
+coupling(cl::Coupling) = cl.θ
+
+"Returns the coupling law constructed from `x`."
+function couple(cl::Coupling, x::AbstractVector)
+    # partition vector using `cl.mask::PartitionMask`
+    x_1, x_2, x_3 = partition(cl.mask, x)
+
+    # construct bijector `B` using θ(x₂)
+    b = cl.θ(x_2)
+
+    return b
+end
+
+function (cl::Coupling)(x::AbstractVector)
+    # partition vector using `cl.mask::PartitionMask`
+    x_1, x_2, x_3 = partition(cl.mask, x)
+
+    # construct bijector `B` using θ(x₂)
+    b = cl.θ(x_2)
+
+    # recombine the vector again using the `PartitionMask`
+    return combine(cl.mask, b(x_1), x_2, x_3)
+end
+(cl::Coupling)(x::AbstractMatrix) = eachcolmaphcat(cl, x)
+
+
+function (icl::Inverse{<:Coupling})(y::AbstractVector)
+    cl = icl.orig
+    
+    y_1, y_2, y_3 = partition(cl.mask, y)
+
+    b = cl.θ(y_2)
+    ib = inv(b)
+
+    return combine(cl.mask, ib(y_1), y_2, y_3)
+end
+(icl::Inverse{<:Coupling})(y::AbstractMatrix) = eachcolmaphcat(icl, y)
+
+function logabsdetjac(cl::Coupling, x::AbstractVector)
+    x_1, x_2, x_3 = partition(cl.mask, x)
+    b = cl.θ(x_2)
+
+    # `B` might be 0-dim in which case it will treat `x_1` as a batch
+    # therefore we sum to ensure such a thing does not happen
+    return sum(logabsdetjac(b, x_1))
+end
+
+function logabsdetjac(cl::Coupling, x::AbstractMatrix)
+    r = [logabsdetjac(cl, x[:, i]) for i = 1:size(x, 2)]
+
+    # FIXME: this really needs to be handled in a better way
+    # We need to return a `TrackedArray`
+    if Tracker.istracked(r[1])
+        return Tracker.collect(r)
+    else
+        return r
+    end
```

Comment:
This is leftovers from me trying to ensure that everything is `TrackedArray` rather than `Array{<:TrackedReal}`. I don't think this is longer something we strive for in Bijectors.jl (e.g. see discussion in #99 ). Sooo maybe just remove this?

Reply:
Thoughts on this @devmotion ?

Reply:
Hmm I think this is ugly, but I know that arrays of TrackedReal are not efficient. However, I guess e.g. if in most cases you just compute the sum of the values then it should be fine?

So if Bijectors in general is not worried about arrays of TrackedReal anymore, then I would be fine without special cases here.

Reply:
Yeah, I remember @mohamed82008 almost preferred having `Array{<:TrackedReal}` because apparently `sum` is faster on this. Though it used to cause a lot of issues with compositions due to the inconsistency, but that was fixed on Tracker's end. We make take a perf hit with compositions though, as I'm guessing 'TrackedArray + TrackedArray' is more efficient than `TrackedArray + Array{<:TrackedReal}` 😕 

I don't know tbh. But this is such an ugly fix anyways, that we might as well just remove it for now 👍